### ISSUE TITLE
adapt_to_samples is slow and only uses one core

### ISSUE THREAD

**Issue by dsmic** [gplepage/vegas#25]
While I was able to use parallelintegrand from the tutorial to speedup point-generation for adapt_to_samples, the adapt_to_samples function is very slow (compared to my very expensive integrand, a some thousands symbol numba compiled sympy expression) .

I could not debug into the function, to get more information, sorry.

I do have about 200 dimensions and generate 200 000 - 2 000 000 points for adaption. Any idea to speed this up?

Thanks Detlef

**Comment by gplepage** [comment]
Sorry for the slow response -- I was on vacation without email.

I will look at it and see what can be done. It could be redesigned to use multiple cores which might help, especially with 200 dimensions (which may be a record for vegas integrands). That would take a while but there may be some other easier fixes I can put in right now.

**Comment by gplepage** [comment]
There isn't much I can do to speed this up other than providing support for multiple cores. The following code defines a standalone function that is equivalent to ``map.adapt_to_sampes(x, fx, nitn, alpha)``  but can use multiple cores:
```python
import vegas
import numpy as np
import multiprocessing

def adapt_to_samples(map, x, fx, nitn=5, alpha=1.0, nproc=1):
    x = np.asarray(x)
    fx = np.asarray(fx)
    map_grid = np.array(map.grid)
    dim = map_grid.shape[0] 
    if dim < nproc:
        nproc = dim 
    args = []
    end = 0
    for i in range(nproc):
        nd = (dim - end) // (nproc - i)
        start = end
        end = start + nd
        args += [(
            map_grid[start:end].flat[:],
            x[:, start:end].flat[:],
            fx,
            nitn, alpha
            )]
    pool = multiprocessing.Pool(processes=nproc)
    results = pool.map(_apply, args)    
    return vegas.AdaptiveMap(np.concatenate(results, axis=0))

def _apply(args):
    grid, x, fx, nitn, alpha = args
    x.shape = (len(fx), -1)
    grid.shape = (x.shape[1], -1)
    map = vegas.AdaptiveMap(grid)
    map.adapt_to_samples(x, fx, nitn=nitn, alpha=alpha)
    return np.asarray(map.grid).tolist() 
```
Use it by setting, for example, ``map = adapt_to_samples(map, x, fx, nitn=nitn, alpha=alpha, nproc=4)`` where 4 cores are used. There is a lot of overhead involved in using multiple cores, of course, so this will be most useful for very high dimensions and/or large numbers of samples. Using 4 cores on my laptop with 200 dimensions and 100,000 samples increases the speed by about a factor of 4.

I need to think about whether I build this into ``vegas`` or not, but you can use the standalone code for now.

**Comment by dsmic** [comment]
Thx, it helps a lot. Time for my quick test with only 69 dimensions 100000 samples went down from 55s to 5s. (12900K 24Threads)

**Comment by gplepage** [comment]
Glad it worked.

Are you running on linux machines?

**Comment by dsmic** [comment]
Yes, Ubuntu 22.04, using a conda environment

**Comment by gplepage** [comment]
Unfortunately the code I suggested above isn't correct. It may provide a map that is useful but it won't be the optimal map. I have updated ``adapt_to_samples`` in the latest version of ``vegas`` (v5.2) so it has an argument ``nproc`` that specifies the number of cores to use. It won't give as big a gain as the code above but it gives the right answer. I get a speedup of about a factor of 3 when using ``nproc=8`` cores for 200 dimensions and 100,000 points.
